packages/grafana-data/src/events/types.ts

```diff
@@ -0,0 +1,109 @@
+import { Unsubscribable, Observable } from 'rxjs';
+
+/*
+ * Event interface
+ */
+export interface BusEvent {
+  readonly type: string;
+  readonly payload?: any;
+}
+
+/*
+ * Base event type
+ */
+export abstract class BusEventBase implements BusEvent {
+  readonly type: string;
+  readonly payload?: any;
+
+  constructor() {
+    //@ts-ignore
+    this.type = this.__proto__.constructor.type;
+  }
+}
+
+/*
+ * Base event type with payload
+ */
+export abstract class BusEventWithPayload<T> extends BusEventBase {
+  readonly payload: T;
+
+  constructor(payload: T) {
+    super();
+    this.payload = payload;
+  }
+}
+
+/*
+ * Interface for an event type constructor
+ */
+export interface BusEventType<T extends BusEvent> {
+  type: string;
+  new (...args: any[]): T;
+}
+
+/*
+ * Event callback/handler type
+ */
+export interface BusEventHandler<T extends BusEvent> {
+  (event: T): void;
+}
+
+/**
+ * Main minimal interface
+ */
+export interface EventBus {
+  /**
+   * Emit single vent
+   */
+  $emit<T extends BusEvent>(event: T): void;
+
+  /**
+   * Subscribe to single event
+   */
+  $on<T extends BusEvent>(eventType: BusEventType<T>, handler: BusEventHandler<T>): Unsubscribable;
+
+  /**
+   * Get observable of events
+   */
+  getStream<T extends BusEvent>(eventType: BusEventType<T>): Observable<T>;
+
+  /**
+   * Remove all event subscriptions
+   */
+  removeAllListeners(): void;
+}
```

Comment:
This is the main new interface, the $ sign is to not conflict with the old methods. The behavior of the handler is a bit different. The handler takes in the full event not just the payload. This will make it possible to subscribe to multiple events with the same handler (in the future).

Reply:
My thinking is that at some point in the future after a looong period of deprecation we can make emit/on functions equal to $emit/$on

Reply:
Would it be possible to rename the $on and $emit to something else? I'm not a big fan of using the $ prefix since it is strongly connected with other libraries and in our case angularjs. When consuming this API it would be nice that it is clear that is has nothing to do with angularjs (if that makes sense?)

Reply:
renamed to publish & subscribe